### PR TITLE
⚡ Implement basic arc-tree building algorithm

### PR DESCRIPTION
**Pull request by paulcscharf**
This is the algorithm that builds the tree for zappy zappy electric arcs.

It's not fully finished, and I'm not sure yet where to put the actual hit-code, but based on our previous discussion, I'd like a review on both the algorithm implementation (to whatever detail you like), as well as your thoughts on where to put that hit-code (see TODO comment).

### CODE REVIEW

File: src/Bearded.TD/Game/Simulation/Elements/ArcTree.cs [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+        GameState game,
+        GameObject? source,
+        IArcSource arcSource,
+        ImmutableArray<Tile> tilesInRange,
+        UntypedDamage damage,
+        TimeSpan lifeTime)
+    {
+        var context = setupContext(game, source, arcSource);
+
+        hitInitialTargets(context, arcSource, tilesInRange);
```
Comment (tomrijnbeek) [inline src/Bearded.TD/Game/Simulation/Elements/ArcTree.cs]: By using the context as object to pass data around, these objects change from pure methods to methods with only side effects. It took me quite long to realise that the method calls below are actually super tied to this method, but there is no obvious data flowing between the two of them. The main culprit is that the context - which is defined as a record - sets the expectation of being immutable. Yes, it uses mutable collections, which strictly speaking - the reference doesn't change - though those always sit in a weird inbetween space when it comes to intuitively determining what is mutability in my opinion.

There are a few solutions I can think of.

The first one is to make the methods actually return something - in this case some form of collection of hits and some form of collection of continuations. That's initializing two collections, yes, but how much of a problem is that really, given that for each continuation we also make a collection of targets using LINQ after getting a collection of tiles? Then you need to move the bookkeeping of the list and queue here, so it creates a bulkier method.

The second one is to move this whole algorithm into a stateful object. You're essentially already dealing with a mutable context object, why not just have:

```cs
var arcPropagation = new ArcPropagation(game, source, arcSource);
arcPropagation.HitInitialTargets(arcSource, tilesInRange);
arcPropagation.PropagateUntilDone();
foreach (var hit in arcPropagation.ResultingArcs)
{
  arc(source, hit);
}
```

This is a pretty "industry standard" solution for complex algorithms like this, though I am not a big fan of it since I'm never sure where to draw the line between making a separate public methods vs just a big `DoTheThing` on it.

The third solution is the least drastic, and that's just to communicate expectations more clearly. Split the context into an immutable part and a mutable part, and call the mutable part something like "accumulator" or "builder".

File: src/Bearded.TD/Game/Simulation/Elements/ArcTree.cs [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+        Context context, int maxCount, ImmutableArray<Tile> tiles)
+    {
+        var targets = tiles
+            .SelectMany(newTargetsOnTileWithPossibleDuplicates())
+            .Distinct()
+            .RandomSubset(maxCount, context.Random);
+
+        return targets;
+
+        Func<Tile, IEnumerable<GameObject>> newTargetsOnTileWithPossibleDuplicates()
```
Comment (tomrijnbeek) [inline src/Bearded.TD/Game/Simulation/Elements/ArcTree.cs]: The func returned is always the same. What is the reason for not using a variable? Or even simpler, make this a local method that takes a `Tile` as argument and returns the `IEnumerable`, so you can use a method reference instead.

File: src/Bearded.TD/Game/Simulation/Elements/ArcTree.cs [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+        {
+            var targets = findTargetsFromContinuation(context, continuation);
+            hitTargets(context, continuation, targets);
+        }
+
+        // TODO: how should damage scale with different hit patterns
+        // variables: bounces, branches, total hits, (distance? nah)
+        // arcSource, as well as modifiers likely want to affect this
+        // could also return a list of hits, and let the caller decide for more flexibility (except YAGNI)
+        foreach (var hit in context.Arcs)
```
Comment (tomrijnbeek) [inline src/Bearded.TD/Game/Simulation/Elements/ArcTree.cs]: Keeping in mind the future where we may want to send the hit tree over the network, it would make sense to make to split this out of this class entirely, and return a constructed arc tree. Then some other algorithm can traverse the tree again and distribute damage (this can be done on the client as well, as it should be deterministic).

File: src/Bearded.TD/Game/Simulation/Elements/ArcTree.cs [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+
+        return context;
+    }
+
+    private static void hitInitialTargets(Context context, IArcSource arcSource, ImmutableArray<Tile> tilesInRange)
+    {
+        var initialContinuation = arcSource.GetInitialArcParametersWithoutModifiers();
+        appleArcModifiers(context, ref initialContinuation);
+
+        var initialTargets = findTargetsInArea(context, initialContinuation.Branches, tilesInRange);
```
Comment (tomrijnbeek) [inline src/Bearded.TD/Game/Simulation/Elements/ArcTree.cs]: You have an `initialContinuation`: is there a reason you can't reuse `findTargetsFromContinuation` here? That would allow you to modify `hitTargets` to not require the `targets` list as parameter any more, since making that list is now shared behaviour.

File: src/Bearded.TD/Game/Simulation/Elements/ArcTree.cs [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+        return targets;
+    }
+
+    private static List<GameObject> findTargetsInArea(
+        Context context, int maxCount, ImmutableArray<Tile> tiles)
+    {
+        var targets = tiles
+            .SelectMany(newTargetsOnTileWithPossibleDuplicates())
+            .Distinct()
+            .RandomSubset(maxCount, context.Random);
```
Comment (tomrijnbeek) [inline src/Bearded.TD/Game/Simulation/Elements/ArcTree.cs]: not actionable: so unfortunate this returns a list and not something immutable ☹️

File: src/Bearded.TD/Game/Simulation/Elements/ArcTree.cs [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+
+    ArcTree.ArcContinuation GetInitialArcParametersWithoutModifiers();
+}
+
+interface IArcModifier
+{
+    void Modify(ref ArcTree.ArcContinuation arc);
+}
+
+static class ArcTree
```
Comment (tomrijnbeek) [inline src/Bearded.TD/Game/Simulation/Elements/ArcTree.cs:27]: Overall, I found it hard to follow the flow of the code in this class. It doesn't help that the nested types are interspersed with the actual methods. It's not uncommon to just skip over nested types since they're not often clever (especially if they're records), but that actually messed me up since understanding that `Context` _is_ actually clever is quite essential to understand the overall flow of code.

File: src/Bearded.TD/Game/Simulation/Elements/ArcTree.cs [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+    private static void hitInitialTargets(Context context, IArcSource arcSource, ImmutableArray<Tile> tilesInRange)
+    {
+        var initialContinuation = arcSource.GetInitialArcParametersWithoutModifiers();
+        appleArcModifiers(context, ref initialContinuation);
+
+        var initialTargets = findTargetsInArea(context, initialContinuation.Branches, tilesInRange);
+        hitTargets(context, initialContinuation, initialTargets);
+    }
+
+    private static void hitTargets(Context context, ArcContinuation previousContinuation, List<GameObject> targets)
```
Comment (tomrijnbeek) [inline src/Bearded.TD/Game/Simulation/Elements/ArcTree.cs]: Since "hitting" is already a very overloaded term in our code (and in general). It seems like what you're actually doing here is "connecting" targets to this tree, isn't that right? Would `connectTargets` or a similar name make it more obvious that this algorithm is basically creating a tree?

**Review by tomrijnbeek — CHANGES_REQUESTED** [review]
I think overall the implementation is pretty good. I can't think of an obvious way to really structure this differently. Complex code like this is notoriously difficult to structure nicely, so I think we'll be hard pressed to find something that is actually simple, since the code just has to have the complexity somewhere.

This is just an initial pass: there is plenty of comments I could place about naming our smaller structures. These higher level comments should already push the code towards more readable, and I'm happy to do further passes if required. Also happy to discuss any of this as well if needed.

File: src/Bearded.TD/Game/Simulation/Elements/ArcTree.cs (hunk 5 above)
Comment (paulcscharf) [reply src/Bearded.TD/Game/Simulation/Elements/ArcTree.cs]: While I understand the sentiment, that happens to be the best internal implementation, and I believe it's generally better to return a specific type rather than a more generic one. Could alternatively work internally with an immutable-array builder, or copy the list into an immutable array, but I don't see the sense in that.

File: src/Bearded.TD/Game/Simulation/Elements/ArcTree.cs (hunk 4 above)
Comment (paulcscharf) [reply src/Bearded.TD/Game/Simulation/Elements/ArcTree.cs]: Took some refactoring, and I don't know if it's strictly speaking better. But I think the difference between the initial and looping parts is clearer now.

File: src/Bearded.TD/Game/Simulation/Elements/ArcTree.cs (hunk 1 above)
Comment (paulcscharf) [reply src/Bearded.TD/Game/Simulation/Elements/ArcTree.cs]: I tried refactoring it in several ways. Ultimately, I prefer keeping the entire state in a single object, which is now called State, implying (I hope) mutability. There are no more methods, so at least there's less magic. It behaves as if this was an object and you'd be calling `this.` instead of `state.` - but I prefer the static methods, rather than having all that code inside the same object as the state.

We could still decide to go this extra step and move the code into the state if we prefer later, and could then even expose it publicly and make it reusable to save on allocations. But I'll worry about that after profiling.

File: src/Bearded.TD/Game/Simulation/Elements/ArcTree.cs (hunk 7 above)
Comment (paulcscharf) [reply src/Bearded.TD/Game/Simulation/Elements/ArcTree.cs]: I like this, making this class explicitly about connecting, and not hitting anything makes it simpler. Someone else can worry about the hitting.

File: src/Bearded.TD/Game/Simulation/Elements/ArcTree.cs [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+        var state = setupState(game, source, arcSource);
+
+        var initialContinuation = createInitialContinuation(state, arcSource);
+        connectToTargetsInArea(state, initialContinuation, areaInRange);
+
+        while (state.QueuedContinuations.TryDequeue(out var continuation))
+        {
+            var area = getAreaForContinuation(state, continuation);
+            connectToTargetsInArea(state, continuation, area);
+        }
```
Comment (paulcscharf) [inline src/Bearded.TD/Game/Simulation/Elements/ArcTree.cs:62]: Compared to before, I think the similarity between these two sections shows more clearly why it can't fully reuse the same code, while clearly highlighting the part it does reuse on a high level instead of hidden somewhere deep in the algorithm.

Comment (paulcscharf) [reply src/Bearded.TD/Game/Simulation/Elements/ArcTree.cs:62]: The only thing that I think isn't quite obvious is that `connectToTargetsInArea` adds further continuations. Or rather: it obviously must be doing so, since that's the recurring thing, but it's maybe not clear why it would exactly. Connecting implies continuing here, and I'm not sure how to represent that better since it happens at the same time in the low-level algorithm below. Calling the method `connectToTargetsInAreaAndQueueContinuations` is possible, but... ugly. Perhaps there's a further refactor that can make that unnecessary or clearer in some other way, but I'm leaving it as is for now.
